File: RemixedDungeon/src/main/java/com/watabou/pixeldungeon/actors/mobs/Succubus.java

package com.watabou.pixeldungeon.actors.mobs;

import com.watabou.pixeldungeon.actors.Char;
import com.watabou.pixeldungeon.actors.CharUtils;
import com.watabou.pixeldungeon.actors.buffs.Buff;
import com.watabou.pixeldungeon.actors.buffs.Charm;
import com.watabou.pixeldungeon.actors.buffs.Sleep;
import com.watabou.pixeldungeon.items.quest.DriedRose;
import com.watabou.pixeldungeon.items.scrolls.ScrollOfLullaby;
import com.watabou.pixeldungeon.items.weapon.enchantments.Leech;
import com.watabou.pixeldungeon.levels.Level;
import com.watabou.utils.Random;

import org.jetbrains.annotations.NotNull;

public class Succubus extends Mob {

	private static final int BLINK_DELAY = 5;

	private int delay = 0;

	public Succubus() {

		hp(ht(80));
		baseDefenseSkill = 25;
		baseAttackSkill  = 40;
		dmgMin = 15;
		dmgMax = 25;
		dr = 10;

		expForKill = 12;
		maxLvl = 25;
		carcassChance = 0;

		loot(new ScrollOfLullaby(), 0.05f);

		addResistance(Leech.class);
		addImmunity(Sleep.class);
	}

	@Override
	public void onSpawn(Level level) {
		super.onSpawn(level);
		setViewDistance(level.getViewDistance() + 1);
	}

	@Override
	public int attackProc(@NotNull Char enemy, int damage) {

		if (Random.Int(3) == 0) {
			Char target = enemy;

			if (enemy.hasBuff(DriedRose.OneWayLoveBuff.class)) {
				target = this;
			}

			float duration = Charm.durationFactor(target) * Random.IntRange(2, 5);

			Buff.affect(target, Charm.class, duration);
		}

		return damage;
	}

	@Override
    public boolean getCloser(int target,  boolean ignorePets) {
		if (level().fieldOfView[target] && level().distance(getPos(), target) > 2 && delay <= 0) {
			CharUtils.blinkTo(this, target);
			delay = BLINK_DELAY;
			spend(-1 / speed());
			return true;
		} else {
			delay--;
			return super.getCloser(target, ignorePets);
		}
	}
}

File: scripts/mobs/Succubus.lua
local RPD = require "scripts/lib/commonClasses"
local mob = require "scripts/lib/mob"

local BLINK_DELAY = 5

-- java Succubus (batch 8): charms on 1/3 hits (OneWayLove redirects the
-- charm to the succubus itself), blinks in from range every 5 turns instead
-- of walking - a free action (spend refund cancels doStepTo's pre-charge).
return mob.init{
    spawn = function(self, level)
        self:setViewDistance(level:getViewDistance() + 1)
    end,

    attackProc = function(self, enemy, dmg)
        if enemy ~= nil and math.random(3) == 1 then
            local target = enemy
            if enemy:hasBuff("OneWayLoveBuff") then
                target = self
            end
            local duration = RPD.Buffs.Charm:durationFactor(target) * math.random(2, 5)
            RPD.affectBuff(target, "Charm", duration)
        end
        return dmg
    end,

    getCloser = function(self, target, ignorePets)
        if self:isPet() then
            return false -- pets use the regular pathfind
        end

        local data = mob.restoreData(self)
        local level = RPD.Dungeon.level

        -- luaj arrays are 1-based: fieldOfView[target + 1] == java fieldOfView[target]
        if level.fieldOfView[target + 1]
                and level:distance(self:getPos(), target) > 2
                and (data.delay or 0) <= 0 then

            local tgtCell = target
            if not level:isCellNonOccupied(target) then
                tgtCell = level:getSafeCellNextTo(target)
            end
            RPD.CharUtils:blinkTo(self, tgtCell)
            data.delay = BLINK_DELAY
            self:spend(-1 / self:speed()) -- refund doStepTo's pre-charge
            return true
        else
            data.delay = (data.delay or 0) - 1
            return false -- regular pathfind
        end
    end
}
